### PR TITLE
chore: fire shutdown signal on anvil node handle drop

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
ref https://github.com/foundry-rs/foundry/pull/8946#pullrequestreview-2324741096

### CODE REVIEW

**Review by mattsse — APPROVED** [review]

File: crates/anvil/src/lib.rs [hunk 1]
```diff
@@ -334,44 +334,51 @@ impl NodeHandle {
...
     pub fn shutdown_signal_mut(&mut self) -> &mut Option<Signal> {
         &mut self._signal
     }
 
-    /// Returns the task manager that can be used to spawn new tasks
+    /// Fires shutdown signal.
+    ///
+    /// Called by long-running tests to make sure anvil instance is terminated.
+    pub fn fire_shutdown_signal(&mut self) {
+        self._signal.take().map(|signal| signal.fire());
```
Comment (DaniPopes) [inline crates/anvil/src/lib.rs:378]: unwrap in the map?

**Review by DaniPopes — COMMENTED** [review]
Q: why is this not called on Drop?

File: crates/anvil/src/lib.rs (hunk 1 above)
Comment (grandizzy) [reply crates/anvil/src/lib.rs]: added in https://github.com/foundry-rs/foundry/pull/8947/commits/654b1590c4cbfb17b78b8a51e79668787bdc0421
also removed testing `DAPP_REMAPPINGS` from env var in cli tests https://github.com/foundry-rs/foundry/pull/8947/commits/98873ec74891ab28bc24341ed7275f6662b8a852, we have test here that uses jail so should be covered https://github.com/foundry-rs/foundry/blob/c9d7b48fb0cdddc33c61db82fc3a94dd7e602c9e/crates/config/src/lib.rs#L3086

**Review by ArkFunds — APPROVED** [review]

**Review by ArkFunds — APPROVED** [review]

File: crates/anvil/src/lib.rs [hunk 2]
```diff
@@ -334,44 +334,53 @@ impl NodeHandle {
...
+    /// This can be used to extract the Signal.
     pub fn shutdown_signal_mut(&mut self) -> &mut Option<Signal> {
         &mut self._signal
     }
 
-    /// Returns the task manager that can be used to spawn new tasks
+    /// Fires shutdown signal.
+    ///
+    /// Called by long-running tests to make sure anvil instance is terminated.
+    pub fn fire_shutdown_signal(&mut self) {
```
Comment (DaniPopes) [inline crates/anvil/src/lib.rs:377]: track_caller

**Review by DaniPopes — APPROVED** [review]

File: crates/anvil/src/lib.rs (hunk 2 above)
Comment (grandizzy) [reply crates/anvil/src/lib.rs:378]: :+1: addeed https://github.com/foundry-rs/foundry/pull/8947/commits/f66e499919d1424df1876f2994e36648eb373479

File: crates/anvil/src/lib.rs [hunk 3]
```diff
@@ -334,44 +334,54 @@ impl NodeHandle {
...
     pub fn shutdown_signal_mut(&mut self) -> &mut Option<Signal> {
         &mut self._signal
     }
 
-    /// Returns the task manager that can be used to spawn new tasks
+    /// Fires shutdown signal.
+    ///
+    /// Called by long-running tests to make sure anvil instance is terminated.
+    #[track_caller]
+    pub fn fire_shutdown_signal(&mut self) {
```
Comment (DaniPopes) [inline crates/anvil/src/lib.rs:378]: Can this go into an `impl Drop`?

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/anvil/src/lib.rs (hunk 3 above)
Comment (grandizzy) [reply crates/anvil/src/lib.rs:378]: yes, will do and make some tests

Comment (grandizzy) [reply crates/anvil/src/lib.rs:378]: hm, weird, I added in drop

```rust
+impl Drop for NodeHandle {
+    fn drop(&mut self) {
+        if let Some(signal) = self._signal.take() {
+            signal.fire().unwrap()
+        }
+    }
+}
```

and then got again a failure like
```
failed to spawn node: Os { code: 98, kind: AddrInUse, message: "Address already in use" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

failures:
    proof::test_account_proof
```

going to make some more tests to see if consistently happens

**Review by ArkFunds — APPROVED** [review]

File: crates/anvil/src/lib.rs (hunk 3 above)
Comment (grandizzy) [reply crates/anvil/src/lib.rs]: added in https://github.com/foundry-rs/foundry/pull/8947/commits/277d62c22a3bcc2484825c3d633517d7df8ac0ca ptal

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**